airbyte-integrations/csv-destination/src/main/java/io/airbyte/integrations/destination/csv/CsvDestination.java

```diff
@@ -0,0 +1,211 @@
+    final String destinationRelativePath = config.get(DESTINATION_PATH_FIELD).asText();
+    Preconditions.checkNotNull(destinationRelativePath);
+
+    return Path.of(destinationRelativePath);
+  }
+
+  /**
+   * This consumer writes individual records to temporary files. If all of the messages are written
+   * successfully, it moves the tmp files to files named by their respective stream. If there are any
+   * failures, nothing is written.
+   */
+  private static class CsvConsumer extends FailureTrackingConsumer<SingerMessage> {
+
+    private final Map<String, WriteConfig> writeConfigs;
+    private final Schema schema;
+
+    public CsvConsumer(Map<String, WriteConfig> writeConfigs, Schema schema) {
+      this.schema = schema;
+      LOGGER.info("initializing consumer.");
+
+      this.writeConfigs = writeConfigs;
+    }
+
+    @Override
+    protected void acceptTracked(SingerMessage singerMessage) throws Exception {
+      if (writeConfigs.containsKey(singerMessage.getStream())) {
+        writeConfigs.get(singerMessage.getStream()).getWriter().printRecord(Jsons.serialize(singerMessage.getRecord()));
+      } else {
+        throw new IllegalArgumentException(
+            String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
+                Jsons.serialize(schema), Jsons.serialize(singerMessage)));
+      }
+    }
+
+    @Override
+    protected void close(boolean hasFailed) throws IOException {
+      LOGGER.info("finalizing consumer.");
+
+      for (final Map.Entry<String, WriteConfig> entries : writeConfigs.entrySet()) {
+        try {
+          entries.getValue().getWriter().flush();
+          entries.getValue().getWriter().close();
+        } catch (Exception e) {
+          hasFailed = true;
+          LOGGER.error("failed to close writer for: {}.", entries.getKey());
+        }
+      }
+      // do not persist the data, if there are any failures.
+      if (!hasFailed) {
+        for (final WriteConfig writeConfig : writeConfigs.values()) {
+          Files.move(writeConfig.getTmpPath(), writeConfig.getFinalPath(), StandardCopyOption.REPLACE_EXISTING);
+        }
+      }
+      // clean up tmp files.
+      for (final WriteConfig writeConfig : writeConfigs.values()) {
+        Files.deleteIfExists(writeConfig.getTmpPath());
+      }
+
+    }
+
+  }
+
+  private static class WriteConfig {
+
+    private final CSVPrinter writer;
+    private final Path tmpPath;
+    private final Path finalPath;
+
+    public WriteConfig(CSVPrinter writer, Path tmpPath, Path finalPath) {
+      this.writer = writer;
+      this.tmpPath = tmpPath;
+      this.finalPath = finalPath;
+    }
+
+    public CSVPrinter getWriter() {
+      return writer;
+    }
+
+    public Path getTmpPath() {
+      return tmpPath;
+    }
+
+    public Path getFinalPath() {
+      return finalPath;
+    }
+
+  }
+
+  public static void main(String[] args) throws Exception {
```

Comment:
Maybe add a comment here so people know this is for local development and testing, not the actual entrypoint to the integration?

Reply:
this changed. this _is_ the actual entrypoint of the destination now. this moved because the version where IntegrationRunner was the entrypoint got into reflection and jar hell.